src/telemetry_poller.erl

```diff
@@ -0,0 +1,373 @@
+%% that the worker is the system's bottleneck. In a healthy situation the message queue length would
+%% be roughly constant.
+%%
+%% === Example - tracking number of active sessions in web application ===
+%%
+%% Let's imagine that you have a web application and you would like to periodically measure number
+%% of active user sessions.
+%%
+%% ```
+%% -module(example_app).
+%%
+%% session_count() ->
+%%    % logic for calculating session count.
+%% '''
+%%
+%% To achieve that, we need a measurement dispatching the value we're interested in:
+%%
+%% ```
+%% -module(example_app_measurements).
+%%
+%% dispatch_session_count() ->
+%%    telemetry:execute([example_app, session_count], example_app:session_count()).
+%% '''
+%%
+%% and tell the Poller to invoke it periodically:
+%%
+%% ```
+%% telemetry_poller:start_link([{measurements, [{example_app_measurements, dispatch_session_count, []}]).
+%% '''
+%%
+%% If you find that you need to somehow label the event values, e.g. differentiate between number of
+%% sessions of regular and admin users, you could use event metadata:
+%%
+%% ```
+%% -module(example_app_measurements).
+%%
+%% dispatch_session_count() ->
+%%    Regulars = example_app:regular_users_session_count(),
+%%    Admins = example_app:admin_users_session_count(),
+%%    telemetry:execute([example_app, session_count], #{count => Admins}, #{role => admin}),
+%%    telemetry:execute([example_app, session_count], #{count => Regulars}, #{role => regular}).
+%% '''
+%%
+%% <blockquote>Note: the other solution would be to dispatch two different events by hooking up
+%% `example_app:regular_users_session_count/0' and `example_app:admin_users_session_count/0'
+%% functions directly. However, if you add more and more user roles to your app, you'll find
+%% yourself creating a new event for each one of them, which will force you to modify existing
+%% event handlers. If you can break down event value by some feature, like user role in this
+%% example, it's usually better to use event metadata than add new events.
+%% </blockquote>
+%%
+%% This is a perfect use case for poller, because you don't need to write a dedicated process
+%% which would call these functions periodically. Additionally, if you find that you need to collect
+%% more statistics like this in the future, you can easily hook them up to the same poller process
+%% and avoid creating lots of processes which would stay idle most of the time.
+%% @end
+%%%---------------------------------------------------
+-module(telemetry_poller).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    child_spec/1,
+    list_measurements/1,
+    start_link/1
+]).
+
+-export([code_change/3, handle_call/3, handle_cast/2,
+     handle_info/2, init/1, terminate/2]).
+
+-define(DEFAULT_VM_MEASUREMENTS, [memory, total_run_queue_lengths]).
+
+-include_lib("kernel/include/logger.hrl").
+-include_lib("kernel/include/logger.hrl").
```

Comment:
This is duplicated and I think `logger` was introduced on OTP 21. We need to add the previous conditionals back. :(